xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
```

Comment:
I inherited from a builtin type because I want the storage as well, not just implementing the interface, since I'd just be proxying a base storage type anyway.

As for order, I said elsewhere: "Note that even if the event system guaranteed order, we're not guaranteed what order the user would load xontribs in, or even what xontribs will be loaded."

Events is a plain object and not a mapping because of syntactic sugar, although there's no reason it can't be a mapping, too.